### PR TITLE
chore: auto generating .tailcallrc.graphql

### PR DESCRIPTION
**Pull request by shashitnak**
**Summary:**  
auto generating the .tailcallrc.graphql file from `Config` type

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [ ] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation] accordingly.
- [x] I have performed a self-review of my code.
- [x] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcall/tree/main/docs

### CODE REVIEW

File: autogen/src/main.rs [hunk 1]
```diff
@@ -76,16 +167,459 @@ async fn update_json() -> Result<()> {
...
+    .metadata
+    .as_ref()
+    .and_then(|metadata| metadata.description.as_ref());
+  write_description(writer, description)?;
+  unsafe {
+    name.as_bytes_mut().get_mut(0).map(|ch| {
+      let lower = (*ch as char).to_ascii_lowercase();
+      *ch = lower as u8;
+    });
+  }
```
Comment (meskill) [inline autogen/src/main.rs:443]: maybe just use some crate instead of playing with unsafe?

File: autogen/src/main.rs [hunk 2]
```diff
@@ -76,16 +167,459 @@ async fn update_json() -> Result<()> {
   Ok(())
 }
 
+fn update_gql() -> Result<()> {
```
Comment (meskill) [inline autogen/src/main.rs:170]: maybe put all the graphql stuff to separate module?

File: autogen/src/main.rs [hunk 3]
```diff
@@ -76,16 +167,459 @@ async fn update_json() -> Result<()> {
...
+
+fn generate_rc_file(file: File) -> Result<()> {
+  let mut file = IndentedWriter::new(file);
+  let mut written_directives = HashSet::new();
+  let wd = &mut written_directives;
+
+  let mut extra_it = BTreeMap::new();
+  let extra_it_ref = &mut extra_it;
+
+  write_schema_for::<config::Server>(&mut file, "Server", "SCHEMA", wd, extra_it_ref)?;
```
Comment (meskill) [inline autogen/src/main.rs:609]: can we use enums instead of string constants?

Comment (shashitnak) [reply autogen/src/main.rs]: converted them to enum

File: autogen/src/main.rs (hunk 2 above)
Comment (shashitnak) [reply autogen/src/main.rs]: moved all the gql code to another file

File: autogen/src/main.rs (hunk 1 above)
Comment (shashitnak) [reply autogen/src/main.rs]: removed unsafe

**Comment by tusharmath** [comment]
**Required**
- [ ] Use https://www.npmjs.com/package/graphql-schema-diff to minimize the diff generated
- [ ] Integrate the script with `lint.sh`

**Comment by shashitnak** [comment]
graphql schema diff based on the current code
BREAKING CHANGES
  ✖ ExprIf was removed.
  ✖ ExprHttp was removed.
  ✖ ExprGrpc was removed.
  ✖ ExprGraphQL was removed.
  ✖ KeyValue was removed.
  ✖ A required field and on input type ExprBody was added.
  ✖ A required field or on input type ExprBody was added.
  ✖ A required field concat on input type ExprBody was added.
  ✖ A required field intersection on input type ExprBody was added.
  ✖ A required field max on input type ExprBody was added.
  ✖ A required field min on input type ExprBody was added.
  ✖ A required field product on input type ExprBody was added.
  ✖ A required field sum on input type ExprBody was added.
  ✖ ExprBody.if changed type from ExprIf to If.
  ✖ ExprBody.http changed type from ExprHttp to Http.
  ✖ ExprBody.grpc changed type from ExprGrpc to Grpc.
  ✖ ExprBody.graphQL changed type from ExprGraphQL to GraphQL.
  ✖ ApplicationXWwwFormUrlEncoded was removed from enum type Encoding.
  ✖ Proxy.url changed type from String to String!.
  ✖ Batch.maxSize changed type from Int to Int!.
  ✖ Batch.delay changed type from Int to Int!.
  ✖ Batch.headers changed type from [String] to [String!]!.
  ✖ groupBy was removed.

**Comment by shashitnak** [comment]
diff for the latest changes
BREAKING CHANGES
  ✖ KeyValue was removed.
  ✖ ExprHttp.headers changed type from [KeyValue] to KeyValues.
  ✖ ExprHttp.query changed type from [KeyValue] to KeyValues.
  ✖ ExprGrpc.headers changed type from [KeyValue] to KeyValues.
  ✖ ExprGrpc.protoPath changed type from String to String!.
  ✖ ExprGraphQL.name changed type from String to String!.
  ✖ ExprGraphQL.args changed type from [KeyValue] to KeyValues.
  ✖ ExprGraphQL.headers changed type from [KeyValue] to KeyValues.
  ✖ ExprGraphQL.batch changed type from Boolean to Boolean!.
  ✖ ApplicationXWwwFormUrlEncoded was removed from enum type Encoding.
  ✖ Proxy.url changed type from String to String!.
  ✖ Batch.maxSize changed type from Int to Int!.
  ✖ Batch.delay changed type from Int to Int!.
  ✖ Batch.headers changed type from [String] to [String!].
  ✖ http was removed.
  ✖ grpc was removed.
  ✖ groupBy was removed.
  ✖ graphQL was removed.

File: autogen/src/gen_gql_schema.rs [hunk 4]
```diff
@@ -0,0 +1,655 @@
...
+    extra_it: &mut BTreeMap<String, ExtraTypes>,
+) -> std::io::Result<()> {
+    let mut nm = reference.split("/").last().unwrap().to_string();
+    first_char_to_upper(&mut nm);
+    extra_it.insert(nm.clone(), ExtraTypes::Schema);
+    write!(writer, "{nm}")
+}
+
+fn first_char_to_upper(name: &mut String) {
+    unsafe {
```
Comment (tusharmath) [inline autogen/src/gen_gql_schema.rs:198]: don't use `unsafe`.

**Review by tusharmath — APPROVED** [review]

**Pull request merged**